File: py/Day07.py
import re
import sys
import time

# Global variables

def generate_result(result, current, remaining_list, part2) -> bool:
    
    if len(remaining_list) > 1:
        a = remaining_list.pop(0)
        b = remaining_list.pop(0)
            
        add = a+b;
        remaining_list.insert(0, add)
        if generate_result(result,add,remaining_list,part2): 
            return True
        
        remaining_list.pop(0)
        
        mul = a*b
        remaining_list.insert(0, mul)
        
        if generate_result(result,mul,remaining_list,part2): 
            return True
        
        remaining_list.pop(0)
        
        if part2: 
            concat = int(f'{a}{b}')
            remaining_list.insert(0, concat)
            if generate_result(result,concat,remaining_list,part2): 
                return True
            
            remaining_list.pop(0)
        
        remaining_list.insert(0,b)
        remaining_list.insert(0,a)
        
    elif len(remaining_list) == 1:
        if result==current: return True
        else: return False
    
    return False

def main():
    global char_array
    
    time_start = int(time.time() * 1000)
    file_path = r'..\input_2024\day07_data.txt'
    
    result_part_1 = 0
    try:
        with open(file_path, "r") as file:
            for line in file:
                result = int(line.split(":")[0])
                numbers = [int(n) for n in line.split(":")[1].split()] 
                if generate_result(result, 0, numbers, False):
                    result_part_1 += result
    except FileNotFoundError:
        print("Error: File not found.")
        return
    print("result part 1:", result_part_1) # correct: 4122618559853
    
    
    result_part_2 = 0
    try:
        with open(file_path, "r") as file:
            for line in file:
                result = int(line.split(":")[0])
                numbers = [int(n) for n in line.split(":")[1].split()] 
                if generate_result(result, 0, numbers, True):
                    result_part_2 += result
    except FileNotFoundError:
        print("Error: File not found.")
        return
    print("result part 2:", result_part_2) # correct: 227615740238334
    
    time_end = int(time.time() * 1000)
    print(f"Execution time: {time_end - time_start} msec") 

if __name__ == "__main__":
    main()
    
    

File: src/AoC2024/Day07.java
package AoC2024;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import AoC2022.Day;

public class Day07 extends Day {

	public Day07() {
		super();
		LOG_LEVEL = 1;
		INPUT_REAL = true;
	}

	private boolean generateResult(long result,Long current,List<Long> remainingList, boolean part2) {
		
		if(remainingList.size() > 1) {
			Long a = remainingList.remove(0);
			Long b = remainingList.remove(0);
			
			Long add = a+b;
			remainingList.add(0, add);
			if(generateResult(result,add,remainingList,part2)) {
				return true;
			}
			remainingList.remove(0);
			
			Long mul = a*b;
			remainingList.add(0, mul);
			
			if(generateResult(result,mul,remainingList,part2)) {
				return true;
			}
			remainingList.remove(0);
			
			if(part2) {
				Long concat = Long.parseLong(String.valueOf(a)+String.valueOf(b));
				remainingList.add(0, concat);
				if(generateResult(result,concat,remainingList,part2)) {
					return true;
				}
				remainingList.remove(0);
			}
			remainingList.add(0,b);
			remainingList.add(0,a);
			
		} else if(remainingList.size() == 1) {
			if(result==current) return true;
			else return false;
		}
		
		return false;
	}
	
	public void processFile() {
		
		long result_part1 = 0;
		String[] lines = sb.toString().split("\n");
		for (String line : lines) {
			long result = Long.parseLong(line.split("\\:")[0]);
			List<Long> numbers =  Arrays.stream(line.split("\\:")[1].trim().split(" ")).map(Long::parseLong).collect(Collectors.toList());
			if(generateResult(result, 0L, numbers,false))
				result_part1 += result;
		}
		logln("result part1: "+result_part1); // correct: 4122618559853
		
		long result_part2 = 0;
		lines = sb.toString().split("\n");
		for (String line : lines) {
			long result = Long.parseLong(line.split("\\:")[0]);
			List<Long> numbers =  Arrays.stream(line.split("\\:")[1].trim().split(" ")).map(Long::parseLong).collect(Collectors.toList());
			if(generateResult(result, 0L, numbers,true))
				result_part2 += result;
		}
		logln("result part2: "+result_part2); // correct: 227615740238334
	}

	public static void main(String[] args) {
		long time_start = System.currentTimeMillis();
		Day07 day = new Day07();
		day.readFile();
		day.processFile();
		long time_end = System.currentTimeMillis();
		day.logln("Execution time: "+(time_end-time_start)+" msec");
		day.flush();
	}


}

